### PR TITLE
Add --test262-esnext option to run-tests.py

### PR DESCRIPTION
**Pull request by ossy-szeged**
Changes:
- Download a quite new test262 revision and run it with the bugfixed test262 harness
- Run tests on CI to be able detect regressions and progressions too

JerryScript-DCO-1.0-Signed-off-by: Csaba Osztrogonác csaba.osztrogonac@h-lab.eu

### CODE REVIEW

File: .github/workflows/gh-actions.yml [hunk 1]
```diff
@@ -74,6 +74,13 @@ jobs:
       - name: Test262 - ES2015
         run: $RUNNER --test262-es2015 update
 
+  Conformance_Tests_ESNext:
+    runs-on: ubuntu-18.04
+    steps:
+      - uses: actions/checkout@v1
+      - name: Test262 - ES2015
```
Comment (rerobika) [inline .github/workflows/gh-actions.yml:81]: ES2015 -> ESNEXT

**Comment by ossy-szeged** [comment]
We agreed that we import the 3rd party test262 harness instead of cloning and patching it. I'm going to update the PR soon.

File: tools/runners/test262-harness.py [hunk 2]
```diff
@@ -0,0 +1,913 @@
...
+
+    def write_log(self, result):
+        name = result.case.get_name()
+        mode = result.case.get_mode()
+        if result.has_unexpected_outcome():
+            if result.case.is_negative():
+                self.logf.write(
+                    "=== %s was expected to fail in %s, but didn't === \n" % (name, mode))
+                self.logf.write("--- expected error: %s ---\n" %
+                                result.case.GetNegativeType())
```
Comment (galpeter) [inline tools/runners/test262-harness.py:846]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 3]
```diff
@@ -0,0 +1,913 @@
...
+        succeeded = progress.succeeded
+        failed = progress.failed
+        write(" - Ran %i test%s" % make_plural(count))
+        if progress.failed == 0:
+            write(" - All tests succeeded")
+        else:
+            write(" - Passed " + percent_format(succeeded, count))
+            write(" - Failed " + percent_format(failed, count))
+            positive = [
+                c for c in progress.failed_tests if not c.case.is_negative()]
```
Comment (galpeter) [inline tools/runners/test262-harness.py:789]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 4]
```diff
@@ -0,0 +1,913 @@
...
+            negative = [
+                c for c in progress.failed_tests if c.case.is_negative()]
+            if positive:
+                print("")
+                write("Failed Tests")
+                for result in positive:
+                    write("  %s in %s" %
+                          (result.case.get_name(), result.case.get_mode()))
```
Comment (galpeter) [inline tools/runners/test262-harness.py:797]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 5]
```diff
@@ -0,0 +1,913 @@
...
+            if negative:
+                print("")
+                write("Expected to fail but passed ---")
+                for result in negative:
+                    write("  %s in %s" %
+                          (result.case.get_name(), result.case.get_mode()))
```
Comment (galpeter) [inline tools/runners/test262-harness.py:803]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 6]
```diff
@@ -0,0 +1,913 @@
...
+                    rel_path = full_path
+                if self.should_run(rel_path, tests):
+                    basename = path.basename(full_path)[:-3]
+                    name = rel_path.split(path.sep)[:-1] + [basename]
+                    if rel_path in self.exclude_list:
+                        print('Excluded: ' + rel_path)
+                    else:
+                        if not self.non_strict_only:
+                            strict_case = TestCase(
+                                self, name, full_path, True)
```
Comment (galpeter) [inline tools/runners/test262-harness.py:755]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 7]
```diff
@@ -0,0 +1,913 @@
...
+                            if not strict_case.is_no_strict():
+                                if strict_case.is_only_strict() or \
+                                        self.unmarked_default in ['both', 'strict']:
+                                    cases.append(strict_case)
+                        if not self.strict_only:
+                            non_strict_case = TestCase(
+                                self, name, full_path, False)
```
Comment (galpeter) [inline tools/runners/test262-harness.py:762]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 8]
```diff
@@ -0,0 +1,913 @@
...
+
+def make_plural(num):
+    if num == 1:
+        return (num, "")
+    return (num, "s")
+
+
+def percent_format(partial, total):
+    return "%i test%s (%.1f%%)" % (make_plural(partial) +
+                                   ((100.0 * partial)/total,))
```
Comment (galpeter) [inline tools/runners/test262-harness.py:673]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 9]
```diff
@@ -0,0 +1,913 @@
...
+
+        if 'raw' in flags:
+            if 'noStrict' in flags:
+                raise TypeError("The `raw` flag implies the `noStrict` flag")
+            elif 'onlyStrict' in flags:
+                raise TypeError(
+                    "The `raw` flag is incompatible with the `onlyStrict` flag")
+            elif self.get_include_list():
+                raise TypeError(
+                    "The `raw` flag is incompatible with the `includes` tag")
```
Comment (galpeter) [inline tools/runners/test262-harness.py:645]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 10]
```diff
@@ -0,0 +1,913 @@
...
+
+        if not flags:
+            return
+
+        if 'raw' in flags:
+            if 'noStrict' in flags:
+                raise TypeError("The `raw` flag implies the `noStrict` flag")
+            elif 'onlyStrict' in flags:
+                raise TypeError(
+                    "The `raw` flag is incompatible with the `onlyStrict` flag")
```
Comment (galpeter) [inline tools/runners/test262-harness.py:642]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 11]
```diff
@@ -0,0 +1,913 @@
...
+
+        if phase not in [None, False, "parse", "early", "runtime", "resolution"]:
+            raise TypeError("Invalid value for negative phase: " + phase)
+
+        if not flags:
+            return
+
+        if 'raw' in flags:
+            if 'noStrict' in flags:
+                raise TypeError("The `raw` flag implies the `noStrict` flag")
```
Comment (galpeter) [inline tools/runners/test262-harness.py:639]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 12]
```diff
@@ -0,0 +1,913 @@
...
+
+    def get_additional_includes(self):
+        return '\n'.join([self.suite.get_include(include) for include in self.get_include_list()])
+
+    def get_source(self):
+        if self.is_raw():
+            return self.test
+
+        source = self.suite.get_include("sta.js") + \
+            self.suite.get_include("assert.js")
```
Comment (galpeter) [inline tools/runners/test262-harness.py:551]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 13]
```diff
@@ -0,0 +1,913 @@
...
+class TestCase(object):
+
+    def __init__(self, suite, name, full_path, strict_mode):
+        self.suite = suite
+        self.name = name
+        self.full_path = full_path
+        self.strict_mode = strict_mode
+        file_desc = open(self.full_path)
+        self.contents = file_desc.read()
+        file_desc.close()
```
Comment (galpeter) [inline tools/runners/test262-harness.py:483]: Please replace this with:
```python
with open(..) as fp:
   self.contents = fp.read()
```

File: tools/runners/test262-harness.py [hunk 14]
```diff
@@ -0,0 +1,913 @@
...
+        self.stderr = stderr
+        self.case = case
+
+    def report_outcome(self, long_format):
+        name = self.case.get_name()
+        mode = self.case.get_mode()
+        if self.has_unexpected_outcome():
+            if self.case.is_negative():
+                print("=== %s was expected to fail in %s, but didn't ===" %
+                      (name, mode))
```
Comment (galpeter) [inline tools/runners/test262-harness.py:428]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 15]
```diff
@@ -0,0 +1,913 @@
...
+                print("--- expected error: %s ---\n" %
+                      self.case.get_negative_type())
```
Comment (galpeter) [inline tools/runners/test262-harness.py:430]: This can be a single line.

File: tools/runners/test262-harness.py [hunk 16]
```diff
@@ -0,0 +1,913 @@
...
+                return True
+        return False
+
+    def get_include(self, name):
+        if not name in self.include_cache:
+            static = path.join(self.lib_root, name)
+            if path.exists(static):
+                file_desc = open(static)
+                contents = file_desc.read()
+                contents = re.sub(r'\r\n', '\n', contents)
```
Comment (galpeter) [inline tools/runners/test262-harness.py:726]: Please use the following style here:
```python
with open(...) as fp:
  contents = fp.read()
contents = re.sub(...)
```

then the `file_desc.close` can be removed from below.

File: tools/runners/test262-harness.py [hunk 17]
```diff
@@ -0,0 +1,913 @@
...
+        logging.basicConfig(level=logging.DEBUG)
+    elif options.loglevel == 'info':
+        logging.basicConfig(level=logging.INFO)
+    elif options.loglevel == 'warning':
+        logging.basicConfig(level=logging.WARNING)
+    elif options.loglevel == 'error':
+        logging.basicConfig(level=logging.ERROR)
+    elif options.loglevel == 'critical':
+        logging.basicConfig(level=logging.CRITICAL)
+    if options.cat:
```
Comment (galpeter) [inline tools/runners/test262-harness.py:896]: Please add a single space before this `if`.

**Pull request merged**